okhttp/src/main/java/okhttp3/HttpUrl.java

```diff
@@ -945,6 +947,50 @@ public static HttpUrl get(URI uri) {
     return url;
   }
 
+  /**
+   * Returns the domain name of this URL's {@link #host()} that is one level beneath the public
+   * suffix by consulting the <a href="https://publicsuffix.org">public suffix list</a>. Returns
+   * null if this URL's {@link #host()} is an IP address or is considered a public suffix by the
+   * public suffix list.
+   *
+   * <p>Use {@link #isPublicSuffix()} to differentiate between private hosts and routable
+   * non-private hosts that might be public suffixes.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>{@code topPrivateDomain()}</th></tr>
+   *   <tr><td>{@code http://google.com}</td><td>{@code "google.com"}</td></tr>
+   *   <tr><td>{@code http://adwords.google.co.uk}</td><td>{@code "google.co.uk"}</td></tr>
+   *   <tr><td>{@code http://square}</td><td>null</td></tr>
+   *   <tr><td>{@code http://co.uk}</td><td>null</td></tr>
+   *   <tr><td>{@code http://localhost}</td><td>null</td></tr>
+   *   <tr><td>{@code http://127.0.0.1}</td><td>null</td></tr>
+   * </table>
+   */
+  public String topPrivateDomain() {
+    if (verifyAsIpAddress(host)) return null;
+    return PublicSuffixDatabase.get().getEffectiveTldPlusOne(host);
+  }
+
+  /**
+   * Returns true if this URL's {@link #host()} is a public suffix contained in the
+   * <a href="https://publicsuffix.org">public suffix list</a>. This will return false if the
+   * {@link #host()} does not explicitly appear in the public suffix list.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>{@code isPublicSuffix()}</th></tr>
+   *   <tr><td>{@code http://com}</td><td>{@code true}</td></tr>
+   *   <tr><td>{@code http://co.uk}</td><td>{@code true}</td></tr>
+   *   <tr><td>{@code http://square}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://google.com}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://localhost}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://127.0.0.1}</td><td>{@code false}</td></tr>
+   * </table>
+   */
+  public boolean isPublicSuffix() {
```

Comment:
My initial reaction to this is mixed. I don’t think it fits as an API on HttpUrl. If you’re doing domain name parsing for something, you’re unlikely to get `https://com/` as a URL.

But I agree that `topPrivateDomain()` is missing one use case, which is that it tells you whether the domain is _private_ but not that it is _routable_. But I also think there’s no way you can discover routability without attempting it. Heck, even `https://com/` could be reachable on some networks. And of course `https://localhost` is reachable.

So – I propose dropping this and focusing the docs in `topPrivateDomain()` on the security/ownership consequences of the name: it’s a unit of trust. Basically we won’t offer any APIs for estimating routability; for that you have to just make the HTTP call.